diff -r 43af55c62d62 external/Tremor-LM-r18595/Makefile.am
--- a/external/Tremor-LM-r18595/Makefile.am	Thu Sep 13 17:07:17 2012 -0300
+++ b/external/Tremor-LM-r18595/Makefile.am	Fri Sep 14 14:54:03 2012 -0300
@@ -25,7 +25,7 @@
 ivorbisfile_example_LDFLAGS = -static 
 ivorbisfile_example_LDADD = libvorbisidec.la
 
-includedir = $(prefix)/include/tremor
+#includedir = $(prefix)/include/tremor
 
 include_HEADERS = ivorbiscodec.h ivorbisfile.h ogg.h os_types.h config_types.h
 
diff -r 43af55c62d62 external/Tremor-r18595/Makefile.am
--- a/external/Tremor-r18595/Makefile.am	Thu Sep 13 17:07:17 2012 -0300
+++ b/external/Tremor-r18595/Makefile.am	Fri Sep 14 14:54:03 2012 -0300
@@ -31,7 +31,7 @@
 iseeking_example_LDFLAGS = -static 
 iseeking_example_LDADD = libvorbisidec.la @OGG_LIBS@
 
-includedir = $(prefix)/include/tremor
+#includedir = $(prefix)/include/tremor
 
 include_HEADERS = ivorbiscodec.h ivorbisfile.h config_types.h
